pkgdb/src/pkgdb/write.cc

```diff
@@ -7,22 +7,293 @@
  *
  * -------------------------------------------------------------------------- */
 
+#include <fstream>
 #include <limits>
 #include <memory>
+#include <optional>
 #include <ranges>
+#include <string>
 
+#include <nlohmann/json.hpp>
+
+#include "flox/core/util.hh"
 #include "flox/flake-package.hh"
 #include "flox/pkgdb/write.hh"
 
 #include "./schemas.hh"
 
-
 /* -------------------------------------------------------------------------- */
 
 namespace flox::pkgdb {
 
 /* -------------------------------------------------------------------------- */
 
+std::string
+scrapeRuleToString( ScrapeRule rule )
+{
+  switch ( rule )
+    {
+      case SR_NONE: return "UNSET";
+      case SR_DEFAULT: return "default";
+      case SR_ALLOW_PACKAGE: return "allowPackage";
+      case SR_DISALLOW_PACKAGE: return "disallowPackage";
+      case SR_ALLOW_RECURSIVE: return "allowRecursive";
+      case SR_DISALLOW_RECURSIVE: return "disallowRecursive";
+      default: return "UNKNOWN";
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+void
+RulesTreeNode::addRule( AttrPathGlob & relPath, ScrapeRule rule )
+{
+  /* Modify our rule. */
+  if ( relPath.empty() )
+    {
+      if ( this->rule != SR_DEFAULT )
+        {
+          // TODO: Pass abs-path
+          throw FloxException( "attempted to overwrite existing rule '"
+                               + scrapeRuleToString( this->rule ) + "' for '"
+                               + this->attrName + "' with new rule '"
+                               + scrapeRuleToString( rule ) + "'" );
+        }
+      traceLog( "assigning rule to `" + scrapeRuleToString( rule ) + "' to `"
```

Comment:
```suggestion
      traceLog( "assigning rule to '" + scrapeRuleToString( rule ) + "' to '"
```
We settled on single quotes across the board for all of our messages. There's a handful of other places this probably needs to be changed.